server/src/telemetry.rs

```diff
@@ -0,0 +1,123 @@
+use restate_core::{cancellation_watcher, TaskCenter, TaskKind};
+use restate_service_client::HttpClient;
+use restate_types::config::CommonOptions;
+use tokio::time::Instant;
+use tracing::debug;
+
+use crate::build_info::{
+    RESTATE_SERVER_DEBUG, RESTATE_SERVER_TARGET_TRIPLE, RESTATE_SERVER_VERSION,
+};
+
+static TELEMETRY_URI: &str = "https://restate.gateway.scarf.sh/restate-server/";
+const TELEMETRY_PERIOD: Duration = Duration::from_secs(3600 * 24);
+
+pub enum Telemetry {
+    Enabled(TelemetryEnabled),
+    Disabled,
+}
+
+pub struct TelemetryEnabled {
+    client: HttpClient,
+    session_id: String,
+    start_time: Instant,
+}
+
+impl Telemetry {
+    pub fn create(options: &CommonOptions) -> Self {
+        if options.disable_telemetry {
+            Self::Disabled
+        } else {
+            let client = HttpClient::from_options(&options.service_client.http);
+            let session_id = ulid::Ulid::new().to_string();
+
+            Self::Enabled(TelemetryEnabled {
+                client,
+                session_id,
+                start_time: Instant::now(),
+            })
+        }
+    }
+
+    pub fn start(self) {
+        match self {
+            Telemetry::Disabled => {}
+            Telemetry::Enabled(enabled) => {
+                if let Err(err) =
+                    TaskCenter::spawn_child(TaskKind::SystemService, "telemetry-service", async {
+                        let cancelled = cancellation_watcher();
+
+                        tokio::select! {
+                            result = enabled.run() => {
+                                result
+                            }
+                            _ = cancelled =>{
+                            }
+                        }
+
+                        Ok(())
+                    })
+                {
+                    debug!(error = %err, "Failed to start telemetry service");
+                }
+            }
+        }
+    }
+}
+
+impl TelemetryEnabled {
+    async fn run(self) {
+        let mut interval = tokio::time::interval(TELEMETRY_PERIOD);
+        loop {
+            interval.tick().await; // first tick completes immediately
+            self.send_telemetry().await;
+        }
+    }
+
+    async fn send_telemetry(&self) {
+        let uptime_hours =
+            (Instant::now().duration_since(self.start_time).as_secs() / 3600).to_string();
+        let session_id = self.session_id.as_str();
+
+        let uri = Uri::from_str(&format!(
+            "{TELEMETRY_URI}?target={RESTATE_SERVER_TARGET_TRIPLE}&version={RESTATE_SERVER_VERSION}&debug={RESTATE_SERVER_DEBUG}&uptime={uptime_hours}&session={session_id}"
+        )).expect("uri must parse");
+
+        debug!(%uri, "Sending telemetry data");
+
+        match self
+            .client
+            .request(
+                uri,
+                None,
+                http::Method::GET,
+                http_body_util::Empty::new(),
+                PathAndQuery::from_static("/"),
```

Comment:
Any specific reason why the query part doesn't go here?

Reply:
the logic in the service client appears to merge the two paths (from the base uri and the path part) and keeps the query from the uri only (because we never expect that paths like /invoke/... or /discover would have a query string). this logic makes sense for how service client is used elsewhere. but here we should be cautious not to use the path part

Reply:
Let's document this behavior. Seems surprising to me.